File: data/languages/cpp/problemtest.cpp
#include "problemtest.h"

#include <algorithm>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <regex>
#include <sstream>
#include <string>
#include <type_traits>
#include <utility>
#include <vector>

#include <nlohmann/json.hpp>

#include "binder.h"
#include "comparator.h"
#include "parser.h"
#include "printer.h"
#include "solutionwrapper.h"

namespace {

bool openFileForWriting(const std::string &filename, std::ofstream &out) {
    std::ofstream outfile(filename, std::ios::out | std::ios::trunc);
    if (!outfile.is_open()) {
        std::cerr << "Unable to open file for writing";
        return false;
    }

    out = std::move(outfile);
    return true;
}

template <typename T>
void print(std::ostream& out, const T& value) {
    out << Printer::toString(value);
}

std::vector<std::string> getTestFiles(const std::string& test_dir_name,
                                      const std::string& testcase_file_name) {
    std::vector<std::string> test_files;

    if (!testcase_file_name.empty()) {
        test_files.emplace_back(testcase_file_name);
        return test_files;
    }

    std::filesystem::directory_iterator dir_iter(test_dir_name);
    for (const auto& entry : dir_iter) {
        if (entry.is_regular_file() && entry.path().extension() == ".test") {
            test_files.emplace_back(entry.path().string());
        }
    }

    // sort such that Test2 comes before Test10
    std::ranges::sort(test_files, [](const std::string& a,
                                     const std::string& b) {
        auto split = [](const std::string& s) {
            std::string before, number, after;
            std::smatch match;
            if (std::regex_search(s, match, std::regex("(\\D*)(\\d*)(\\D*)"))) {
                before = match[1].str();
                number = match[2].str();
                after = match[3].str();
            }
            return std::make_tuple(before, number.empty() ? -1 : std::stoi(number), after);
        };
        const auto split_a = split(a);
        const auto split_b = split(b);

        if (std::get<1>(split_a) != -1 && std::get<1>(split_b) != -1) {
            return split(a) < split(b);
        }
        return a < b;
    });
    
    return test_files;
}

std::vector<std::string> toLines(const std::string& testcase_file) {
    std::vector<std::string> lines;
    std::ifstream in(testcase_file, std::ios::in);
    std::string line;
    while (std::getline(in, line)) {
        lines.emplace_back(std::move(line));
    }
    return lines;
}

}  // annoymous namespace

ProblemTest::ProblemTest(const std::string& test_dir_name,
                         const std::string& results_file_name,
                         const std::string& testcase_file_name)
    : test_dir_name_(test_dir_name),
      results_file_name_(results_file_name),
      testcase_file_name_(testcase_file_name) {
}

auto getDurationSince(const auto& start) {
    const auto end = std::chrono::high_resolution_clock::now();
    return std::chrono::duration_cast<std::chrono::milliseconds>(
        end - start).count();
}

bool ProblemTest::runTest(
    const std::string& testcase_file_name, nlohmann::json& test) const {
    bool success = false;
    bool areEqual = false;
    Solution solution;

    Binder::return_type ret{};
    Binder::return_type expected{};

    test["testcase_name"] = std::filesystem::path(testcase_file_name).stem();

    try {
        auto lines = toLines(testcase_file_name);
        constexpr size_t expected_testcase_size =
            Binder::func_arg_size_v + 1;
        if (expected_testcase_size != lines.size()) {
            std::cerr << "Incorrect number of parameters specified in the"
                    << " testcase file. Must specify one line per "
                    << "parameter + one line for expected output. "
                    << "Found: " << lines.size() << " Expected: "
                    << expected_testcase_size << "." << std::endl;
            success = false;
        } else {
            string expected_str;
            std::swap(expected_str, lines.back());
            expected = std::move(parse<Binder::return_type>(expected_str));
            ret = std::move(Binder::solve(solution, std::move(lines)));
            success = Comparator::compare(ret, expected, true);
            if (!success) {
                test["reason"] =
                    "Mismatch between expected and actual output.";
            }
        }
    } catch (const std::exception& e) {
        std::stringstream ss;
        std::cerr << "Exception occurred while running the test case. "
                  << "Exception: " << std::string(e.what());
        test["reason"] = "Exception occurred.";
        success = false;
    } catch (...) {
        test["reason"] = "Exception occurred.";
        std::cerr << "Unknown exception occurred while running the test case.";
        success = false;
    }

    if (success) {
        test["status"] = "Success";
    } else {
        test["status"] = "Failed";
        test["expected"] = expected;
        test["actual"] = ret;
        test["testcase_file"] = testcase_file_name;
    }

    return success;
}

bool ProblemTest::run() const {
    const auto start = std::chrono::high_resolution_clock::now();
    bool success = false;
    std::vector<nlohmann::json> testsJason;
    
    const vector<std::string> test_files = getTestFiles(test_dir_name_,
                                                        testcase_file_name_);

    for (const auto& testcase_file : test_files) {
        testsJason.emplace_back();
        success = runTest(testcase_file, testsJason.back());
        if (!success) {
            break;
        }
    }

    nlohmann::json jsonObj;
    jsonObj["duration_ms"] = getDurationSince(start);

    if (success) {
        jsonObj["status"] = "Ok";
        jsonObj["tests"] = testsJason;
    } else {
        jsonObj["status"] = "Failed";
        jsonObj["tests"] = testsJason;
    }

    std::ofstream out;
    if (!openFileForWriting(results_file_name_, out)) {
        std::cerr << "Error: Could not open output file: "
                  << results_file_name_
                  << std::endl;
        return 1;
    }
    out << std::setw(4) << jsonObj << std::endl;

    return true;
}

File: data/languages/python/problemtest.py
import importlib.util
import inspect
import json
import os
import re
import sys
import time

from typing import get_args
from typing import get_origin

from treenode import TreeNode

ANY_VALUE = "*"


def split_top_level(value, delimiter):
    tokens = []
    token = []
    depth = 0
    in_string = False
    escape = False

    for char in value:
        if escape:
            token.append(char)
            escape = False
            continue

        if char == "\\" and in_string:
            token.append(char)
            escape = True
            continue

        if char == "\"":
            in_string = not in_string
            token.append(char)
            continue

        if not in_string:
            if char == "[":
                depth += 1
            elif char == "]":
                depth -= 1
            elif char == delimiter and depth == 0:
                tokens.append("".join(token).strip())
                token = []
                continue

        token.append(char)

    if token or value == "":
        tokens.append("".join(token).strip())
    return [token for token in tokens if token != ""]


def parse_bool(value):
    if value == "true":
        return True
    if value == "false":
        return False
    raise ValueError(f"Invalid boolean format: {value}")


def parse_string(value):
    if len(value) < 2 or value[0] != "\"" or value[-1] != "\"":
        raise ValueError(f"Invalid string format: {value}")
    return bytes(value[1:-1], "utf-8").decode("unicode_escape")


def parse_array(value, element_type):
    stripped = value.strip()
    if len(stripped) < 2 or stripped[0] != "[" or stripped[-1] != "]":
        raise ValueError(f"Invalid array format: {value}")
    inner = stripped[1:-1].strip()
    if inner == "":
        return []
    return [parse_typed_value(token, element_type)
            for token in split_top_level(inner, ",")]


def build_tree(values):
    if not values or values[0] is None:
        return None

    root = TreeNode(values[0])
    queue = [root]
    index = 1

    while queue and index < len(values):
        node = queue.pop(0)

        if index < len(values) and values[index] is not None:
            node.left = TreeNode(values[index])
            queue.append(node.left)
        index += 1

        if index < len(values) and values[index] is not None:
            node.right = TreeNode(values[index])
            queue.append(node.right)
        index += 1

    return root


def parse_tree(value):
    stripped = value.strip()
    if stripped == "null":
        return None
    if len(stripped) < 2 or stripped[0] != "[" or stripped[-1] != "]":
        raise ValueError(f"Invalid tree format: {value}")

    inner = stripped[1:-1].strip()
    if inner == "":
        return None

    values = []
    for token in split_top_level(inner, ","):
        token = token.strip()
        if token == "null":
            values.append(None)
        else:
            values.append(int(token))
    return build_tree(values)


def parse_typed_value(value, annotation):
    value = value.strip()

    if annotation is inspect._empty:
        raise ValueError("All Python solution parameters and returns need type hints.")

    if annotation is bool:
        return parse_bool(value)
    if annotation is int:
        return int(value)
    if annotation is str:
        return parse_string(value)
    if annotation is TreeNode:
        return parse_tree(value)

    origin = get_origin(annotation)
    args = get_args(annotation)
    if origin is list:
        if len(args) != 1:
            raise ValueError(f"Unsupported list annotation: {annotation}")
        return parse_array(value, args[0])

    raise ValueError(f"Unsupported annotation: {annotation}")


def serialize_tree(root):
    if root is None:
        return []

    result = []
    queue = [root]
    while queue:
        node = queue.pop(0)
        if node is None:
            result.append(None)
            continue
        result.append(node.val)
        queue.append(node.left)
        queue.append(node.right)

    while result and result[-1] is None:
        result.pop()
    return result


def serialize_value(value):
    if value is None:
        return "None"
    if isinstance(value, TreeNode):
        return serialize_tree(value)
    if isinstance(value, list):
        return [serialize_value(item) for item in value]
    return value


def compare_values(actual, expected):
    return serialize_value(actual) == serialize_value(expected)


def get_testcase_name(filename):
    base = os.path.basename(filename)
    if base.endswith(".test"):
        return base[:-5]
    return base


def read_lines(filename):
    with open(filename, "r", encoding="utf-8") as handle:
        return [line.rstrip("\n") for line in handle]


def get_test_files(test_dir_name, testcase_name):
    if testcase_name.lower() != "all":
        filename = os.path.join(test_dir_name, testcase_name + ".test")
        if not os.path.isfile(filename):
            raise FileNotFoundError(f"Testcase file does not exist: {filename}")
        return [filename]

    test_files = []
    for entry in os.listdir(test_dir_name):
        if entry.endswith(".test"):
            test_files.append(os.path.join(test_dir_name, entry))

    def sort_key(path):
        return [
            int(text) if text.isdigit() else text.lower()
            for text in re.split(r"([0-9]+)", os.path.basename(path))
        ]

    return sorted(test_files, key=sort_key)


def load_solution_class():
    expected = os.environ.get("OPENLEETCODE_EXPECTED") == "1"
    filename = "solution_expected.py" if expected else "solution.py"
    filepath = os.path.join(os.getcwd(), filename)

    spec = importlib.util.spec_from_file_location("openleetcode_solution", filepath)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"Could not load solution module from {filepath}")

    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)

    solution_class = getattr(module, "Solution", None)
    if solution_class is None:
        raise RuntimeError(f"Solution class not found in {filename}")
    return solution_class


def get_solution_method(solution):
    methods = []
    for name, method in inspect.getmembers(solution, predicate=callable):
        if name.startswith("_"):
            continue
        if getattr(method, "__self__", None) is not solution:
            continue
        methods.append((name, method))

    if len(methods) != 1:
        raise RuntimeError("Solution must define exactly one public method.")
    return methods[0]


def run_test(solution_class, testcase_file_name, test):
    solution = solution_class()
    method_name, method = get_solution_method(solution)
    signature = inspect.signature(method)

    test["testcase_name"] = get_testcase_name(testcase_file_name)

    lines = read_lines(testcase_file_name)
    expected_testcase_size = len(signature.parameters) + 1
    if len(lines) != expected_testcase_size:
        reason = (
            "Incorrect number of parameters specified in the testcase file. "
            "Must specify one line per parameter + one line for expected output. "
            f"Found: {len(lines)} Expected: {expected_testcase_size}."
        )
        test["status"] = "Failed"
        test["reason"] = reason
        test["expected"] = ""
        test["actual"] = ""
        test["testcase_file"] = testcase_file_name
        return False

    expected_str = lines[-1]
    arg_lines = lines[:-1]
    parameters = list(signature.parameters.values())
    parsed_args = [
        parse_typed_value(arg_lines[index], parameter.annotation)
        for index, parameter in enumerate(parameters)
    ]

    actual = method(*parsed_args)
    if expected_str == ANY_VALUE:
        test["status"] = "Skipped"
        test["actual"] = serialize_value(actual)
        return True

    expected = parse_typed_value(expected_str, signature.return_annotation)
    if compare_values(actual, expected):
        test["status"] = "Success"
        return True

    test["status"] = "Failed"
    test["reason"] = "Mismatch between expected and actual output."
    test["expected"] = serialize_value(expected)
    test["actual"] = serialize_value(actual)
    test["testcase_file"] = testcase_file_name
    return False


def main():
    if len(sys.argv) < 4:
        print(
            f"Usage: {sys.argv[0]} <testcase directory> <results output file> "
            "<testcase name>",
            file=sys.stderr,
        )
        return 1

    test_dir_name = sys.argv[1]
    results_file_name = sys.argv[2]
    testcase_name = sys.argv[3]

    start = time.time()
    tests_json = []
    success = True

    try:
        solution_class = load_solution_class()
        test_files = get_test_files(test_dir_name, testcase_name)
        for testcase_file in test_files:
            test_json = {}
            tests_json.append(test_json)
            success = run_test(solution_class, testcase_file, test_json)
            if not success:
                break
    except Exception as error:
        tests_json.append({
            "status": "Failed",
            "testcase_name": testcase_name,
            "reason": "Exception occurred.",
            "expected": "",
            "actual": "",
            "testcase_file": test_dir_name,
        })
        print(str(error), file=sys.stderr)
        success = False

    json_obj = {
        "duration_ms": int((time.time() - start) * 1000),
        "testcase_filter_name": testcase_name,
        "status": "Ok" if success else "Failed",
        "tests": tests_json,
    }

    with open(results_file_name, "w", encoding="utf-8") as handle:
        json.dump(json_obj, handle, indent=4)

    return 0 if success else 1


if __name__ == "__main__":
    raise SystemExit(main())
